platforms/extensibility/plugin-use/src/main/java/org/gradle/plugin/use/internal/DefaultPluginRequestApplicator.java

```diff
@@ -226,41 +169,54 @@ private PluginResolver wrapInAlreadyInClasspathResolver(ClassLoaderScope classLo
+        private final PluginResolution resolved;
+
+        public ApplyAction(PluginRequestInternal request, PluginResolution resolved) {
+            this.request = request;
+            this.resolved = resolved;
+        }
+
+        public void apply(PluginManagerInternal target) {
             try {
-                applicator.run();
-                pluginApplicationListenerBroadcaster.pluginApplied(request);
-            } catch (UnknownPluginException e) {
-                throw couldNotApply(request, id, e);
+                try {
+                    pluginApplicationListenerBroadcaster.pluginApplied(request);
+                    resolved.applyTo(target);
+                } catch (UnknownPluginException e) {
+                    throw couldNotApply(request, request.getId(), e);
+                } catch (Exception e) {
+                    throw exceptionOccurred(request, e);
+                }
             } catch (Exception e) {
-                throw exceptionOccurred(request, e);
+                throw new LocationAwareException(e, request.getScriptDisplayName(), request.getLineNumber());
             }
-        } catch (Exception e) {
-            throw new LocationAwareException(e, request.getScriptDisplayName(), request.getLineNumber());
         }
     }
 
     private static boolean noPluginsApplied(PluginRequests requests, PluginRequests autoAppliedPlugins) {
         return requests.isEmpty() && autoAppliedPlugins.isEmpty();
     }
 
-    private InvalidPluginException couldNotApply(PluginRequestInternal request, PluginId id, UnknownPluginException cause) {
+    private static InvalidPluginException couldNotApply(PluginRequestInternal request, PluginId id, UnknownPluginException cause) {
         return new InvalidPluginException(
             String.format(
                 "Could not apply requested plugin %s as it does not provide a plugin with id '%s'."
                     + " This is caused by an incorrect plugin implementation."
                     + " Please contact the plugin author(s).",
-                request, id),
+                request.getDisplayName(), id),
             cause);
     }
 
-    private InvalidPluginException exceptionOccurred(PluginRequestInternal request, Exception e) {
-        return new InvalidPluginException(String.format("An exception occurred applying plugin request %s", request), e);
+    private static InvalidPluginException exceptionOccurred(PluginRequestInternal request, Exception e) {
+        return new InvalidPluginException(String.format("An exception occurred applying plugin request %s", request.getDisplayName()), e);
     }
 
-    private Result resolveToFoundResult(PluginResolver effectivePluginResolver, PluginRequestInternal request) {
-        Result result = new Result(request);
+    private static Result resolveToFoundResult(PluginResolver effectivePluginResolver, PluginRequestInternal request) {
+        Result result = new Result();
```

Comment:
💭 Is there a good reason why `PluginResolver` takes a result as an argument and mutates it?  This seems like the only implementation of `PluginResolutionResult`.  I wonder if it would be cleaner to have `PluginResolver.resolve()` return a `PluginResolutionResult` and have found/notFound just be static methods on `PluginResolutionResult` that return an instance of that interface? (it looks like those methods only ever get called once for any given plugin request, but maybe I missed something)

Reply:
I imagine it is twofold. One, this method makes it a bit simpler to collect multiple _notFound_ results, so we can print all the places we _tried_ to resolve a plugin, if no plugin was found. Two, this is older code and immutability was less cool back then. 

I've updated this to an immutable `PluginResolutionResult` implementation

Reply:
The methods do get called more than once, as there are resolvers that delegate to other resolvers, some of which are composite resolvers.